python_modules/dagster/dagster/_core/definitions/asset_dep.py

```diff
@@ -49,13 +53,10 @@ class AssetDep(
 
     def __new__(
         cls,
-        asset: Union[CoercibleToAssetKey, AssetSpec, AssetsDefinition, SourceAsset],
+        asset: CoercibleToAssetDep,
         partition_mapping: Optional[PartitionMapping] = None,
     ):
-        if isinstance(asset, AssetSpec):
-            asset_key = asset.asset_key
-        else:
-            asset_key = AssetKey.from_coercible_or_definition(asset)
+        asset_key = AssetKey.from_coercible_to_asset_dep(asset)
```

Comment:
In the case when `asset` is already an `AssetDep`, we have four options:
1. create a new `AssetDep` and discard the `partition_mapping` already on `asset` in favor of whatever is passed as `partition_mapping` (even if it is `None`)
2. create a new `AssetDep` using the `partition_mapping` already on `asset`, even if `partition_mapping` is provided 
3. Error if `asset.partition_mapping` and `partition_mapping` are both not None and not the same partition mapping 
4. Error, full stop. State that users should not be creating `AssetDep`s from existing `AssetDep`s

Reply:
I think it's reasonable to claim that users shouldn't be passing `AssetDep`s to `AssetDep(...)`, and then relax the constraint if we get asked to. However, I don't like that the type checking is done at runtime. We could instead manually define the type allowed as `Union[CoercibleToAssetKey, AssetSpec, AssetsDefinition, SourceAsset]` to that editors will catch the error

Reply:
I don't think we should allow `AssetDep` here as the behavior of the `partition_mapping` will always be compromised and strange.

Reply:
ok. does keeping one long union type for the `__new__` seem ok then? 

```python
def __new__(
        cls,
        asset:  Union[CoercibleToAssetKey, AssetSpec, AssetsDefinition, SourceAsset],
        partition_mapping: Optional[PartitionMapping] = None,
    ):
```

the alternate would be to include more type aliases